flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java

```diff
@@ -535,41 +545,82 @@ protected void beforeInvoke() throws Exception {
         // -------- Invoke --------
         LOG.debug("Invoking {}", getName());
 
+        SuspendableMailboxLoop suspendableLoop = mailboxProcessor.getSuspendableLoop();
+
         // we need to make sure that any triggers scheduled in open() cannot be
         // executed before all operators are opened
-        actionExecutor.runThrowing(
-                () -> {
-                    SequentialChannelStateReader reader =
-                            getEnvironment()
-                                    .getTaskStateManager()
-                                    .getSequentialChannelStateReader();
-                    reader.readOutputData(
-                            getEnvironment().getAllWriters(),
-                            !configuration.isGraphContainingLoops());
-
-                    operatorChain.initializeStateAndOpenOperators(
-                            createStreamTaskStateInitializer());
-
-                    channelIOExecutor.execute(
-                            () -> {
-                                try {
-                                    reader.readInputData(getEnvironment().getAllInputGates());
-                                } catch (Exception e) {
-                                    asyncExceptionHandler.handleAsyncException(
-                                            "Unable to read channel state", e);
-                                }
-                            });
+        CompletableFuture<Void> allGatesRecoveredFuture =
+                actionExecutor.call(
+                        () -> {
```

Comment:
nit: is it feasible to extract this lambda function to a separate method? `restore()` has grown a bit too much.  Or maybe portions of it?

Reply:
It is exactly what I want to do. In fact, I already did it in a separate commit but I didn't push it yet.